File: internal/model/stack.go
package model

import "xompose/internal/model/entity"

type StackModel struct {
	entity.Stack
}

type StackItem struct {
	Id          int         `json:"id"`
	StackName   string      `json:"stackName"`
	YamlPath    string      `json:"yamlPath"`
	YamlContent string      `json:"yamlContent"`
	EnvContent  string      `json:"envContent"`
	StackStatus StackStatus `json:"stackStatus"`
}

type StackAddReq struct {
	StackName   string `json:"stackName"`
	YamlContent string `json:"yamlContent"`
	Envs        []map[string]any
}

type StackGetReq struct {
	StackName string `v:"required#请输入 stack 名称"`
}

type StackListReq struct {
	// PageReq
}

type StackStatus string

const (
	StackRunning StackStatus = "running"
	StackCreated StackStatus = "created"
)

type StackListItem struct {
	*StackItem
	IsManaged bool `json:"isManaged"`
}

func (m *StackModel) ToItem() *StackItem {
	return &StackItem{
		Id:          m.Id,
		StackName:   m.StackName,
		StackStatus: StackStatus(m.StackStatus),
		YamlPath:    m.YamlContent,
		YamlContent: m.YamlContent,
		EnvContent:  m.EnvContent,
	}
}

type StackDeleteReq struct {
	StackName string `json:"stackName"`
}

type StackStatusItem struct {
	Image   string      `json:"image"`
	Service string      `json:"service"`
	State   StackStatus `json:"state"`
	Ports   []*Port     `json:"ports"`
}

type Port struct {
	PublishPort int    `json:"publishPort"`
	TargetPort  int    `json:"targetPort"`
	Protocol    string `json:"protocol"`
}

File: web/src/types/stack.ts

export type StackList = {
    id: string
    name: string
    tags: string[]
}

export type StackListResp = {
    items: Array<ListItem>
}

export type ListItem = StackItem & {
    isManaged: Boolean
}

export type StackItem = {
    id?: string
    stackName: string
    stackStatus?: string
    yamlContent: string
    yamlPath: string
    envContent: string
}


export const UNKNOWN = 0
export const CREATED_FILE = 1
export const CREATED_STACK = 2
export const RUNNING = 3
export const EXITED = 4


export const statusColor = (status: number): string => {
    switch (status) {
        case CREATED_FILE:
            return "dark";
        case CREATED_STACK:
            return "dark";
        case RUNNING:
            return "primary";
        case EXITED:
            return "danger";
        default:
            return "secondary";
    }
}

export const statusNameShort = (status: number): string => {
    switch (status) {
        case CREATED_FILE:
            return "inactive";
        case CREATED_STACK:
            return "inactive";
        case RUNNING:
            return "active";
        case EXITED:
            return "exited";
        default:
            return "?";
    }
}

export type StackStatusListResp = {
    items: StackStatusItem[]
}

export type StackStatusItem = {
    service: string
    image: string
    state: string
    ports: Port[]
    networks: string[]
    volumes: string[]
    dependsOn: string[]
}

export type Port = {
    publishPort: number
    targetPort : number
    protocol: string
}